lib/linked_list.rb

```diff
@@ -12,150 +12,225 @@ def initialize(value, next_node = nil)
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_first
-      raise NotImplementedError
-    end
-
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def add_last(value)
-      raise NotImplementedError
-    end
-
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_last
-      raise NotImplementedError
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+  
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def add_first(value)
+    @head = Node.new(value, @head)
+    return @head
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def search(value)
+    return false if @head.nil?
+    
+    current = @head
+    
+    while !current.nil?
+      if current.data == value
+        return true
+      end
+      current = current.next
     end
+    false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    return nil if @head.nil?
+    current = @head
+    max = current.data
+    while !current.nil? do
+      max = current.data if max < current.data
+      current = current.next 
+    end
+    return max
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_min
+    return nil if @head.nil?
+
+    current = @head
+    min = current.data
+
+    while !current.nil?
+      min = current.data if min > current.data
+      current = current.next 
+    end
+    return min
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def length
+    current = @head
+    length = 0
+
+    return 0 if current.nil?
+    
+    while !current.nil?
+      current = current.next
+      length += 1
+    end
+    return length
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    return nil if @head.nil?
+    
+    current = @head
+
+    index.times do
+      return nil if current.next.nil?
+      current = current.next
+    end
+    
+    return current.data
+  end
+  
+  # method to print all the values in the linked list
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def visit

```

Comment:
This method should be printing out all the values of the linked list.  You're doing something odd here:

This is assigning current to it's data and if that's truthy, you're then crashing the program by trying to access `current.data` because `current` has been assigned to `current.data` previously.

```ruby
if current = current.data
  return current.data
end
```